### ISSUE TITLE
Select set of builds for initial mandatory auto PR testing process

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#2317]
**CC:** @trilinos/framework 

## Description

This story is to involve interested members of the Trilinos team to collaborate to select the best build configurations in order to allow the new Trilinos auto PR testing and merging tool and process (#1155) to become the mandatory way to test and push to the main Trilinos ‘develop’ branch (#2312).  This selection must be done considering that there are limited computing resources (on the Jenkins build farms) to run automated builds.  So while we would like to run many different useful builds as part of pre-push automated PR testing, we have to be strategic about what builds we run where to not overwhelm current capacity. As more build machines are added to the Jenkins build farm, more builds can be added to the auto PR testing process.

This story is a follow on from the action item:

* ACTION (Ross): Set up a separate meeting to discuss what that build / those builds (if more than one) should be

in the [2018-02-26 Trilinos Planning Meeting](https://docs.google.com/document/d/1JClLSR3n79XJT_yLPPoQv_e3rJToHxl8fUzRYojtY5Y/edit#bookmark=id.9wrhu8wvpblx).

The set of Trilinos team members interested in being part of this discussion (and meetings) include:

* @mhoemmen 
* @crtrott 
* @william76
* @bartlettroscoe
* @rppawlo
* @ibaned
* @jwillenbring
* @bmpersc

Since this selection of these builds will impact every Trilinos developer and every close customer and collaboration of Trilinos, it is important that we get input from many different people in making this selection.

## Definition of Done

* Document conversation between Trilinos developers on the selection of these builds
* New build configurations selected (with actual Trilinos configuration files)
* Trial builds of Trilinos posted to CDash for the chosen configurations

## Related Issues

* Related to: #1154 
* Part of: #2312 
* Composed of: #2462, #2463, #2464

## Task

1. Find initial selection of team members interested to discuss this topic **[DONE]**
1. Set up and have meeting with working group **[DONE]**
1. Create initial set of builds in meeting **[DONE]**
1. Create concrete *`.cmake` files for each proposed configuration and set up nightly builds sumitting to "Specialized" CDash Track/Group:
  a. GCC 4.8.4: See #2462 
  b. Intel 17.x: See #2463
  c. CUDA: See #2464
1. ???

**Comment by bartlettroscoe** [comment]
I sent the following email.  I will give it until the week of 3/12/2-18 after the SIAM PP conference to have this meeting.  That should be urgent enough.

**Comment by csiefer2** [comment]
Any machine used for PR testing must be accessible (either directly or by an identical system) by all developers so they can fix issues that PR testing exposes.

**Comment by mhoemmen** [comment]
I'm with @csiefer2 -- for example, if we intend to support Windows builds, the Windows Dashboard build should be a VM that we can access (either log into, or download and run on our own) and use for testing.

**Comment by bartlettroscoe** [comment]
I don't think there is no way that we would make a Windows build a auto PR build or even a CI build anytime soon.  There are so many other builds that more important than that.

But I think the rule should be that before any build can enter the "Nightly", "Clean", or "Auto PR" category, that it must be easy for any Sandia staff member to access a machine where the build can be exactly reproduced.  I agree with @csiefer2 and @mhoemmen on this point.

**Comment by bartlettroscoe** [comment]
FYI: @trilinos/muelu 

It appears that the current auto PR testing GCC 4.8.4 posting to CDash never runs any MueLu tests as shown at:

* https://testing-vm.sandia.gov/cdash/index.php?project=Trilinos&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=3&showfilters=1&filtercombine=and&field1=groupname&compare1=61&value1=Pull%20Request&field2=buildstarttime&compare2=84&value2=now&field3=subprojects&compare3=93&value3=MueLu

Not sure why that is because the standard CI build set up for last 1.5 has been running MueLu tests without any issues (except for real failures) as shown in recent history at:

* https://testing-vm.sandia.gov/cdash/index.php?project=Trilinos&filtercombine=and&filtercombine=and&filtercount=4&showfilters=1&filtercombine=and&field1=buildname&compare1=66&value1=-MPI_RELEASE_DEBUG_SHARED_PT_CI&field2=groupname&compare2=61&value2=Continuous&field3=buildstarttime&compare3=84&value3=now&field4=subprojects&compare4=93&value4=MueLu

**Comment by bartlettroscoe** [comment]
FYI: @trilinos/anasazi 

It also appears that the auto PR builds are not running any Anazazi tests either as shown by:

* https://testing-vm.sandia.gov/cdash/index.php?project=Trilinos&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=3&showfilters=1&filtercombine=and&field1=groupname&compare1=61&value1=Pull%20Request&field2=buildstarttime&compare2=84&value2=now&field3=subprojects&compare3=93&value3=Anasazi

**Comment by bartlettroscoe** [comment]
FYI: @trilinos/ifpack2 

It also appears that the auto R builds are not running any Ifpack tests as shown by:

* https://testing-vm.sandia.gov/cdash/index.php?project=Trilinos&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=3&showfilters=1&filtercombine=and&field1=groupname&compare1=61&value1=Pull%20Request&field2=buildstarttime&compare2=84&value2=now&field3=subprojects&compare3=93&value3=Ifpack2

**Comment by jwillenbring** [comment]
@bartlettroscoe Thank you for the reminder about the MueLu, Ifpack2, and Anasazi tests. These packages had test failures a while back and were disabled for that reason (needed 100% passing). Some of these issues I believe have been resolved. I am testing that now.

**Comment by bartlettroscoe** [comment]
> Thank you for the reminder about the MueLu, Ifpack2, and Anasazi tests. These packages had test failures a while back and were disabled for that reason (needed 100% passing). Some of these issues I believe have been resolved. I am testing that now.

@jwillenbring,

Just the failing tests should be disabled, not all the tests for a package.  You write a GitHub issue for the failing test(s), then disable just those tests.  Generally one wants to disable with a scapulae, not a machete.

**Comment by jwillenbring** [comment]
> Just the failing tests should be disabled, not all the tests for a package. You write a GitHub issue for the failing test(s), then disable just those tests. Generally one wants to disable with a scapulae, not a machete.

Agreed. At the time I didn't want the issues to be lost and not notice that the tests were disabled. Much of this has to do with ETI.

**Comment by bartlettroscoe** [comment]
> At the time I didn't want the issues to be lost and not notice that the tests were disabled. Much of this has to do with ETI.

If the developers of those packages don't care to fix the broken tests for GitHub issues associated with their packages then that is on them.  And you only disable the tests just for that particular build, not for all builds.  Look at the git log file for `cmake/std/BasicCiTestingSettings.cmake` to see how to do this.  These tests will still show up failing in "Clean", "Nightly" and other builds but at least they will not mess up other developers PR testing.  That way, no one will forget these failed tests.  But if they let the tests fail for more than a day or two, then these tests need to be disabled in "Clean" and "Nightly" as well.

It is important to fix this quickly because developers might be thinking that the auto PR builds are testing Ifpack2, Anasasi, and MueLu but they are not.  This means that the auto PR builds have even a higher chance of breaking tests in these packages.

**Comment by mhoemmen** [comment]
Wait, what?!?  Ifpack2 and MueLu tests were totally disabled?!?

**Comment by bartlettroscoe** [comment]
Hello, it is after SIAM PP.  Can we set up a meeting on this?

**Comment by bartlettroscoe** [comment]
@jwillenbring and/or @bmpersc,

Do you guys want to be included in this meeting?  If so, please add your GitHub IDs to the list in the [above](https://github.com/trilinos/Trilinos/issues/2317#issue-301582151) Description field.  I will then try to set up a meeting using people's SNL calendars to select a time that works for everyone.  (Or if I can't find a time, I will set up a dreaded doodle.com poll.)

**Comment by bartlettroscoe** [comment]
One issue to consider is that according to:

* https://gcc.gnu.org/wiki/openmp

GCC 4.8.4 should implement OpenMP 3.1 while GCC 4.9.3 should implement OpenMP 4.0.

Is that an issue for testing OpenMP code with Trilinos?  How much difference is there in the Kokkos implementation of threading of OpenMP 3.1 vs OpenMP 4.0?  Or is this not a concern because updates to Kokkos do testing with many different compilers before merging to Trilinos 'develop'?

If we are not worried about the OpenMP 3.1 vs. OpenMP 4.0 issue, then a single GCC 4.8.4 build for auto PR testing would seem to be okay.

**Comment by nmhamster** [comment]
@bartlettroscoe - OpenMP3.1 would be good for now because we have a range of platforms where this is the most up to date complete supported standard.

**Comment by ndellingwood** [comment]
@bartlettroscoe each package in each of ShyLU's subpackages is marked either ST or EX:

[shylu_dd](https://github.com/trilinos/Trilinos/blob/3fbb278edb8ec7323f59928d750ab0dd0a49ec30/packages/shylu/shylu_dd/cmake/Dependencies.cmake#L5)

[shylu_node](https://github.com/trilinos/Trilinos/blob/3fbb278edb8ec7323f59928d750ab0dd0a49ec30/packages/shylu/shylu_node/cmake/Dependencies.cmake#L5)

I'm not sure why they are being listed as PT packages, the use of 
```
TRIBITS_PACKAGE_DEFINE_DEPENDENCIES(
  SUBPACKAGES_DIRS_CLASSIFICATIONS_OPTREQS
...
```
in the `shylu_node` and `shylu_dd` subpackages seems to fit the example from [tribits dev guide example](https://tribits.org/doc/TribitsDevelopersGuide.html#tribitsexampleproject). Does anything stand out that clearly needs to be changed? 

Opening a new issue after speaking with @srajama1 to get packages migrated to PT testing, will cross reference.

**Comment by bartlettroscoe** [comment]
NOTE: Defects like #2391 show why we need to be enabling OpenMP in our auto PR build.

**Comment by bartlettroscoe** [comment]
**CC:** @trilinos/zoltan2

I was just reminded today by #2397 that the auto PR builds based on SEMS should also enable the Scotch TPL.  That CI build (and the builds derived from it) are the only automated builds of Trilinos that enable Scotch and therefore are the only builds that run these tests that depend on Scotch.

**Comment by bartlettroscoe** [comment]
Got feedback from @bathmatt  and @micahahoward on the builds they would like to see to help protect ATDM APP SPARC and EMPIRE usage of Trilinos.  The below email thread provides a converssation.  In short, they would like to see following builds tested in the auto PR tester (in ATDM [Trilinos build keywords](https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/ATDM+Builds+of+Trilinos) terms):

* **gnu-debug-serial** (GCC compilers, CMAKE_BUILD_TYPE=DEBUG, Kokkos Node = Serial)
* **intel-opt-openmp** (Intel compilers, CMAKE_BUILD_TYPE=RELEASE, Kokkos Node = OpenMP)
* **cuda-opt** (GCC + CUDA, CMAKE_BUILD_TYPE=RELEASE, Kokkos Node = CUDA)

@bathmatt  and @micahahoward,

Would it be just a good if gnu and intel switched and we added debug-mode checking to the CMAKE_BUILD_TYPE=RELEASE to catch more undefined memory usage problems with:

* **gnu-opt-openmp** (Intel compilers, CMAKE_BUILD_TYPE=RELEASE, Trilinos_ENABLE_DEBUG=ON, Kokkos Node = OpenMP)
* **intel-debug-serial** (Intel compilers, CMAKE_BUILD_TYPE=DEBUG, Kokkos Node = Serial)
* **cuda-opt** (GCC + CUDA, CMAKE_BUILD_TYPE=RELEASE, Kokkos Node = CUDA)

?

That would allow people to run with debug-mode checking and OpenMP on machines that did not have an Intel compiler, or there was an Intel license sever problem (which happens from time to time).  Is there something better about testing OpenMP with Intel than with GCC?

Also, is testing with a serial node (i.e. no Kokkos threading) really an important use case to test pre-merge to 'develop'?  Is this a debugging tool in case something goes wrong with the OpenMP Kokkos node, even with `OMP_NUM_THREADS=1`?  If OpenMP works, why would you not use that build for development, debugging, or running problems?

<details>
<summary>
  <b>Detailed email thread</b> (Click to expand)
</summary>

From: Howard, Micah 
Sent: Wednesday, March 21, 2018 2:17 PM
To: Bartlett, Roscoe A
Subject: Re: mandatory PR testing..

Yes that's okay with me.

----

From: "Bartlett, Roscoe A"
Date: Wednesday, March 21, 2018 at 11:58 AM
To: "Howard, Micah", "Bettencourt, Matthew"
Subject: RE: mandatory PR testing..

Micah and Matt,
 
The nightly ATDM builds of Trilinos we are setting up will cover all of these (and will completely cover them for ATDM once we include the stuff needed for SPARC as well).  The issue is what can we afford to run for PR testing prior to merge.  Once that becomes the mandatory way to test and push to Trilinos 'develop', there will be a lot of active PR branches at a time and that is going to likely require 10-20 builds on some days.  We can manage that on the current Ted Bed or other hardware for that many different build configurations.
 
The reduced list of three builds below may be doable for PR testing before updating the 'develop' branch.  But there may not be enough computational resources to run even these three builds.  I will ask the various ATDM DevOps supporting teams where we can find stable reliable resources to run these builds.  (FYI: 'white' and 'ride' seem pretty unstable right now for running larger jobs with 'bsub'.)  But no one seems to know how loaded these machines really are and there are no tools set up to really tell you that.
 
After cleaning up a little, is okay If I archive this email chain in:
 
       https://github.com/trilinos/Trilinos/issues/2317
 
?
 
I think it would be good to get you guys on record with what you think will be valuable for ATDM.
 
Thanks,
 
-Ross

---
 
From: Howard, Micah 
Sent: Wednesday, March 21, 2018 1:50 PM
To: Bettencourt, Matthew ; Bartlett, Roscoe A
Subject: Re: mandatory PR testing..
 
Sounds reasonable.
 
For PR acceptance, note that:
    DEBUG-gcc-serial -> is my (4) with DEBUG
    REL-icc-openmp -> is my (2) with REL
    rel-cuda -> is my (3) with REL
 
And arguably my (1) and (4) are redundant.

---

From: "Bettencourt, Matthew"
Date: Wednesday, March 21, 2018 at 11:45 AM
To: "Howard, Micah", "Bartlett, Roscoe A"
Subject: Re: mandatory PR testing..
 
I think we may want this level of testing for dev->master, my only concern with this level of testing is that it will be a long time between PR and acceptance into develop.  
My ideal situation is that we have a more rapid push into develop so garbage doesn't end up there, and an automated nightly upgrade to master based on Micah's suggestion.  With this I'd use master for most everything, develop for co-develop work or short term bug fix testing (like the bug Roger found today) on my desktop and we could have rapid turn around.  Therefore anything that blocks promotion from devel->master would be all hands on deck moment, things that break PR should be handled at that level.

----

On 03/21/2018 11:40 AM, Howard, Micah wrote:

I sent this to Ross when I declined his meeting for tomorrow (I have a conflict too)

1) A surrogate for Trinity/HSW : intel-17.0.4, mpich2-3.2, Kokkos/Serial
2) A surrogate for Trinity/KNL: intel-17.04, mpich2-3.2, Kokkos/OpenMP
3) A surrogate for Sierra(ATS-2): gcc-7.2.0, openmpi-2.1 (I think), Kokkos/CUDA-9.x
3b) I'd like to have coverage for IBM's XL compiler and MPI along with CUDA-9.x but this dev stack is not stable yet
4) A surrogate for a 'standard' Linux env (this encompasses CEE, TLCC2 & CTS-1 systems): gcc or clang, openmpi, Kokkos/Serial, specific versions of gcc or clang and openmpi TBD

So very similar to what Matt suggested.  I'd add that REL and DEBUG are both useful for these cases.  Complete testing with DEBUG for all cases might be asking a bit much but ensure DEBUG builds aren't broken is useful.

Micah

---

On 3/21/18, 11:26 AM, "Bettencourt, Matthew" wrote:

Ross,

I've got a conflict for you meeting on thur, I'd suggest

DEBUG-gcc-serial
REL-icc-openmp
rel-cuda

At least 1+3 since they are very different,

</details>

**Comment by nmhamster** [comment]
@bartlettroscoe - running with OpenMP enabled but intended to do MPI-only (i.e. `OMP_NUM_THREADS=1`) has been show to add fairly serious performance overhead because of reduced compiler optimization and lower efficiency (the OpenMP runtime is outlining functions which are not needed in the MPI only runs). For systems like ATS1/Haswell partition, we would strongly recommend MPI only (i.e. `Serial`) as the fastest configuration.

@bartlettroscoe - in general if we use a sufficiently new GCC, the difference in supported OpenMP features is fairly low. Using older GCC means that Intel would be a better choice.

**Comment by bartlettroscoe** [comment]
@nmhamster,

> For systems like ATS1/Haswell partition, we would strongly recommend MPI only (i.e. Serial) as the fastest configuration.

Okay, so testing with a serial non-threaded Kokkos node is important to test.  Is the compiler in that case always going to be some variant of GNU?

> in general if we use a sufficiently new GCC, the difference in supported OpenMP features is fairly low. 

I think GCC used for PR testing needs to be GCC 4.8.4, because that is the minimum required version of GCC for Kokkos.  Is that considered "old" from an OpenMP perspective compared to say Intel 15.0..2 (which I think is the minimum version of Intel that Kokkos currently supports).

So which is better:

* **gnu-debug-serial** (GCC 4.8.4 compilers, CMAKE_BUILD_TYPE=DEBUG, Kokkos Node = Serial)
* **intel-opt-openmp** (Intel 15.0.2 compilers, CMAKE_BUILD_TYPE=RELEASE, Kokkos Node = OpenMP)

or

* **gnu-opt-openmp** (GCC 4.8.4 compilers, CMAKE_BUILD_TYPE=RELEASE, Trilinos_ENABLE_DEBUG=ON, Kokkos Node = OpenMP)
* **intel-debug-serial** (Intel 15.0.2 compilers, CMAKE_BUILD_TYPE=DEBUG, Kokkos Node = Serial)

from this ATDM perspective?

**Comment by nmhamster** [comment]
@bartlettroscoe 

> Okay, so testing with a serial non-threaded Kokkos node is important to test. Is the compiler in that case always going to be some variant of GNU?

More likely to be Intel on ATS1 or CTS platforms. Rarely use GNU in production environments.

> I think GCC used for PR testing needs to be GCC 4.8.4, because that is the minimum required version of GCC for Kokkos. Is that considered "old" from an OpenMP perspective compared to say Intel 15.0..2 (which I think is the minimum version of Intel that Kokkos currently supports).

Both are quite old but they have a fairly consistent set of OpenMP features. The more useful case is Intel 17.4 (with GCC 4.9.3 header files) because this is the current configuration used for production builds. Obviously, over time this will change but updates are infrequent.

**Comment by nmhamster** [comment]
@bartlettroscoe - the Intel comment relates to ATS1, for ATS2 then GCC is more likely. So its quite complicated as we move forward and have to support many platforms. Sorry to be a pain.

**Comment by bartlettroscoe** [comment]
@nmhamster,

> More likely to be Intel on ATS1 or CTS platforms. Rarely use GNU in production environments.

So that means that `intel-opt-serial` is a more useful build to help support ATDM than `gnu-opt-serial`, right?

Remember, this is just the auto PR builds, not all of the builds.  We are setting up automated ATDM-focused builds of Trilinos on every system and with every configuration that is important to ATDM.  It is up to us in ATDM to decide what those builds should be.

**Comment by nmhamster** [comment]
@bartlettroscoe 

> So that means that intel-opt-serial is a more useful build to help support ATDM than gnu-opt-serial, right?

Yes, I think this would be more useful.

> Remember, this is just the auto PR builds, not all of the builds. We are setting up automated ATDM-focused builds of Trilinos on every system and with every configuration that is important to ATDM. It is up to us in ATDM to decide what those builds should be.

Understood.

**Comment by micahahoward** [comment]
Not sure I can offer much that hasn't already been said.  Let me know if you need anything from me.

**Comment by mhoemmen** [comment]
@bartlettroscoe The only issue with CUDA debug testing is that it's really REALLY slow (e.g., every `Kokkos::View` array access gets checked on the GPU), so you may need to increase the time-out.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe The only issue with CUDA debug testing is that it's really REALLY slow (e.g., every Kokkos::View array access gets checked on the GPU), so you may need to increase the time-out. 

@mhoemmen, or just perhaps only run the faster `BASIC` test suite?  But I think the CUDA build should be `cuda-opt`, not `cuda-debug`.  ATDM is running full `cuda-debug` and there are just a few issues with Panzer tests timing out (see #2318).

**Comment by mhoemmen** [comment]
@bartlettroscoe That's cool -- just note that the slowness comes from bounds checking and other Kokkos debug features, not from `-g -O0`.  Thus, "release build with bounds checking on" will be nearly as slow as "debug build with bounds checking on."

**Comment by bartlettroscoe** [comment]
We had a meeting last Thursday to discuss the selection of the auto PR builds which involved the people @bartlettroscoe, @mhoemmen, @ibaned, @bmpersc, @crtrott, @jwillenbring, @rppawlo, @agsalin, and @william76.  The builds we selected are shown in the Google Doc page:

* https://docs.google.com/document/d/1bCqOJLoXSnq24ct-IjwGkvse6xhsp9HrHsxqZxBwDWQ

In sujmary, we selected the following builds:

* SEMS GCC 4.8.4, OpenMPI 1.8.7, Primary Tested Packages and TPLs, `CMAKE_BUILD_TYPE=RELEASE`, `Trilinos_ENABLE_DEBUG=ON`, `OpenMP` (`OMP_NUM_THREADS=2`)

* SEMS Intel 17.x, OpenMPI 2.x, Primary Tested Packages and TPLs, `CMAKE_BUILD_TYPE=RELEASE`, `Trilinos_ENABLE_DEBUG=OFF`, 

* CUDA (will match the ATDM CUDA build on 'white', see [ATDM Builds of Trilinos](https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/ATDM+Builds+of+Trilinos) )

All of the builds have the common options:
* `BUILD_SHARED_LIBS=ON`
* `Trilinos_ENABLE_EXPLICIT_TEMPLATE_INSTANTIATION=ON`
* `Xpetra_ENABLE_Experimental=ON`
* `MueLu_ENABLE_Experimental=ON`
* `Trilinos_TRACE_ADD_TEST=ON`
* `Trilinos_TEST_CATEGORIES=BASIC`

I will create some new GitHub issues for each of these target builds to outline what needs to be done to get them ready to plug into the auto PR tester.

**Comment by csiefer2** [comment]
I am uncomfortable testing with these options enabled:

Xpetra_ENABLE_Experimental=ON
MueLu_ENABLE_Experimental=ON

We're still waiting on an explanation of why the apps want those...

**Comment by bartlettroscoe** [comment]
> I am uncomfortable testing with these options enabled:
> 
> Xpetra_ENABLE_Experimental=ON
> MueLu_ENABLE_Experimental=ON
> 
> We're still waiting on an explanation of why the apps want those...

That is a negotiation that needs to take place between MueLu developer, ATDM APP developers and management.  But as long as these are getting enabled in ATDM builds of Trilinos, we need to enable them in these builds as well to protect ATDM customers.

**Comment by csiefer2** [comment]
Well consider my objection logged then :)

**Comment by mhoemmen** [comment]
@csiefer2 If it helps, I objected strenuously (alas, I don't think it will help, this sounds like one of those "SURPRISE WE'RE USING THIS NOW YOU'D BETTER SUPPORT IT" things).

**Comment by csiefer2** [comment]
@mhoemmen Awww, thanks!

**Comment by tawiesn** [comment]
@csiefer2  @mhoemmen To make you feel more comfortable: it helps to grep through the Xpetra and MueLu source code for the Experimental flag.
- Xpetra: the Experimental flag only guards the code in the research folder (which is basically nothing). All other functionality (including the support for blocked operators, Thyra, etc...) is not experimental any more (i removed it months ago). We did not remove the flag as CMake option, though, to allow people to work on new experimental stuff without having to add a new Experimental flag. Furthermore, removing a CMake variable produces annoying warnings for people who are still trying to use them.
- MueLu: there is not much guarded by the Experimental flag: it's basically the Matlab interface (@csiefer2: who needs that Matlab interface? ;-) ), the AMGX interface (has it ever been used?) and some minor stuff (Stratimikos). I'm very sure one could quickly check what is really needed and get rid of the Experimental flag in the corresponding places. Something that could be done in minutes rather than hours...

Happy easter holidays and greetings from Europe!

**Comment by csiefer2** [comment]
@tawiesn : Matlab is not experimental.  That's TPL guaded.

**Comment by tawiesn** [comment]
@csiefer2 That's right. It's marked to be experimental only in the ReleaseNotes.txt file. Even better!

**Comment by mhoemmen** [comment]
@tawiesn Thanks and happy Easter to you too!  I'm not sure then why ATDM needs that flag.  I would feel better if the ATDM build did not enable it, or if we could identify exactly what they need and make just that not experimental.

**Comment by tawiesn** [comment]
@mhoemmen 

> if we could identify exactly what they need and make just that not experimental.

I think that would be the right thing to do. And i am very sure it's not much work. But you have definitely more time/resources and the ability to run the tests than i have with my little tiny 8 year old computer.

If i have to make a guess: it's probably something in the MueLu Stratimikos interface which we use for years without problems and therefore can safely mark as non-experimental.

**Comment by tawiesn** [comment]
@mhoemmen @csiefer2 
> #documentationisalwaysuptodate

Clean code principle: Don’t document your code. Code your documentation.

Believe it or not: that really works.

**Comment by bartlettroscoe** [comment]
FYI: @micahahoward last week told me that he things that using OpenMP with the Intel 17.x build would be better than with the GCC 4.8.4 build from an ATDM perspective.

Does anyone have an objection to making that change to the plan for auto PR builds?

**Comment by bartlettroscoe** [comment]
The EMPIRE build of Trilinos has removed the experimental Xpetra and MueLu enables and these have been removed from the ATDM and the CI build (see commits 7481c76 and 71c51f2).  I updated the document:

* https://docs.google.com/document/d/1bCqOJLoXSnq24ct-IjwGkvse6xhsp9HrHsxqZxBwDWQ

to remove these options.

**Comment by mhoemmen** [comment]
@prwolfe @micahahoward Hooray! :D

**Comment by bartlettroscoe** [comment]
@crtrott and @nmhamster,

Since we can't turn on OpenMP in our auto PR builds yet due to the problem of pinning to the same cores as described in #2422, would a short-term alternative be to try to use Pthreads for threading with Kokkos?  That is, use Pthread for the Kokkos host threading model?

Also, why does Kokkos not automatically use Pthread for threading when the TPL Pthread is enabled and no other threading model is specified (and OpenMP, CUDA, etc.) are not enabled or specified?  You ca see that simply enabling the Pthread TPL does not turn on Kokkos usage of Pthread as shown in the recent auto PR build:

* https://testing-vm.sandia.gov/cdash/viewConfigure.php?buildid=3444702

which shows:

```
Explicitly enabled TPLs on input (by user):  Pthread MPI BLAS LAPACK Boost ParMETIS Zlib HDF5 Netcdf SuperLU BoostLib DLlib 12

...

-- Setting KokkosCore_ENABLE_Pthread=ON since TPL_ENABLE_Pthread=ON

...

-- ****************** Kokkos Settings ******************
-- Execution Spaces
--   Device Parallel: None
--     Host Parallel: None
--       Host Serial: Serial
-- 
-- Architectures:
--     None
-- 
-- Enabled options
--   KOKKOS_ENABLE_SERIAL
--   KOKKOS_ENABLE_DEBUG
--   KOKKOS_ENABLE_PROFILING
--   KOKKOS_ENABLE_DEPRECATED_CODE

...
```

Therefore, enabling the Pthread TPL in this build does nothing, right?

**Comment by mhoemmen** [comment]
@bartlettroscoe wrote:
> would a short-term alternative be to try to use Pthreads for threading with Kokkos?

Kokkos folks have been telling everyone not to use the Pthreads back-end for quite a while.  

> Also, why does Kokkos not automatically use Pthread for threading when the TPL Pthread is enabled and no other threading model is specified (and OpenMP, CUDA, etc.) are not enabled or specified?

The Pthread TPL gets automatically detected and enabled by default.  (There are reasons to enable it that have nothing to do with thread parallelism.)  Users found it unpleasantly surprising to get thread parallelism when they didn't ask for it explicitly.  Thus, Kokkos and Tpetra historically did not enable the Pthreads back-end unless explicitly requested.  Tpetra treats `Kokkos::Threads` as a last resort in terms of defaults.

More importantly, OpenMP is the only back-end that promises polite sharing of hardware with other libraries that use OpenMP.  The Pthreads back-end does not share hardware resources nicely with other libraries that use threads (even Pthreads).

**Comment by mhoemmen** [comment]
@bartlettroscoe There is a bit of a fun history.  "Kokkos 1.0" (Chris Baker's thread-parallel programming model) had the following back-ends originally:

  - TBB (Intel's Threading Building Blocks), called TBBNode
  - ThreadPool (Carter Edwards' Pthreads-based thread pool), called TPINode
  - CUDA
  - Serial (no threads)

The "OpenMPNode" came later and was actually an external contribution.  Historically TBBNode didn't get exercised so much, because it required a TPL.  CUDA was very poorly tested.  This made TPINode a common thing for people to try.  None of the Nodes performed particularly well in Tpetra until I specialized some of the Serial kernels.  We eventually replaced Kokkos 1.0 with Kokkos 2.0 in Tpetra; the fact that Nodes still exist is mainly because there is no schedule for Trilinos 13.0 (we've given up & have plans to get rid of them by making them unnecessary).

**Comment by ibaned** [comment]
@bartlettroscoe although @dsunder has the most expertise in this, I don't think the Threads backend (soon to be powered by `std::thread` rather than `pthread`) has an advantage over OpenMP in terms of behaving well when not bound (it may be better to use OpenMP with `OMP_PROC_BIND` specifically set to not bind). For practical ATDM concerns, OpenMP is probably their main backend of interest and I don't know of a case where ATDM cares about the Threads backend.

@mhoemmen there is some cool history there, I didn't know that. Also I was hoping there was a schedule for Trilinos 13.0, we were planning to line that up with Kokkos 3.0.

**Comment by mhoemmen** [comment]
@ibaned At one point there was a plan for quarterly minor releases & yearly major releases.  That all fell apart :( .  It's hard to respect the backwards compatibility plan (deprecate before the last minor release, remove at the next major release) when there's no schedule for releases.

**Comment by mhoemmen** [comment]
@ibaned wrote:
> Also I was hoping there was a schedule for Trilinos 13.0, we were planning to line that up with Kokkos 3.0.

I haven't heard anything about a 13.0 release, but that doesn't necessarily mean it won't happen :)

**Comment by ibaned** [comment]
@bartlettroscoe enabling the Pthread TPL and (only) the Kokkos OpenMP backend should be just fine, I think thats what currently happens in most Trilinos builds that enable OpenMP.

**Comment by bartlettroscoe** [comment]
> Edit: the most common configuration enables the Kokkos OpenMP and Serial backends.

@ibaned, does this mean that we should or should not allow the enable of the Pthread TPL when we configure Trilinos with OpenMP or Serial Kokkos backends?

@mhoemmen, other than some than some tests that need `std::` C++11 threading to run, what other reasons are there for enabling the Pthread TPL when configuring Trilinos, even if Kokkos will use a different (or no) threading backend?

**Comment by ibaned** [comment]
The Pthread TPL is fine to enable regardless of Kokkos backends. It doesn't cause any issues with OpenMP or otherwise. It is also fine to disable, unless the Kokkos Threads backend is used.

**Comment by mhoemmen** [comment]
@bartlettroscoe wrote:
> ... what other reasons are there for enabling the Pthread TPL when configuring Trilinos, even if Kokkos will use a different (or no) threading backend?

I actually searched Trilinos for `pthread_` and found nothing.  Teuchos requires C++11 so we should all use the C++ Standard Library stuff like `std::call_once` and `std::mutex` if we need thread synchronization.

**Comment by mhoemmen** [comment]
@bartlettroscoe My above comments suggest that maybe we don't need the Pthread TPL at all.  I've been fine with it enabled but perhaps it's not necessary.  On the other hand, sometimes OpenMP implementations or other TPLs need it, so I'm not sure we should just turn it off by default.

**Comment by bartlettroscoe** [comment]
> My above comments suggest that maybe we don't need the Pthread TPL at all. I've been fine with it enabled but perhaps it's not necessary. On the other hand, sometimes OpenMP implementations or other TPLs need it, so I'm not sure we should just turn it off by default.

@mhoemmen, the experience with the Teuchos MM classes thread-safe work with that you need to explicitly link in the `-lpthread` lib if you want to use C++11 threading support.  This is needed to run the  multi-threaded unit tests.

**Comment by bartlettroscoe** [comment]
FYI: As pointed out by @etphipp in https://github.com/trilinos/Trilinos/issues/2628#issuecomment-384347891, setting:

```
 -D MPI_EXEC_PRE_NUMPROCS_FLAGS="--bind-to;none"
```

seems to fix the problem of OpenMP threads all binding to the same core on a RHEL6 machine.  Perhaps this will let us enable OpenMP for the GCC 4.8.4 auto PR build being set up in #2462?

**Comment by jwillenbring** [comment]
@bartlettroscoe Since the PR builds have been running for some time I propose we close this ticket and deal with additional issues in other tickets.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe Since the PR builds have been running for some time I propose we close this ticket and deal with additional issues in other tickets.

@jwillenbring, sure, the only remaining build is the CUDA build in #2464 and we have a plan for that (see the issue).

Closing.
